DLPJTS-110 automatic calculation

datalogics-kam opened the pull request:
**Note: Please do [datalogics/pdf-java-toolkit#1007](https://octocat.dlogics.com/datalogics/pdf-java-toolkit/pull/1007) first!**

#### Changes in this pull request

- Remove explicit call to running calculate scripts.
- Update input and output files accordingly.

#### Fulfills  [DLPJTS-110](https://jira.datalogics.com/browse/DLPJTS-110)

#### Checklist for approving this pull request

(PR creator amend this with more conditions if necessary)

- [x] There are **unit tests** for new/changed code, or a good explanation why this was not possible.
- [x] All **CI builders** have indicated success (Give them a few minutes to notice the pull request.)
- [x] The **Pull request title** has the JIRA issue numbers separated by spaces (if any), a space, and then a short, but descriptive summary.
- [x] **Commit messages** are well formed: [A note about Git commit messages](http://www.tpope.net/node/106)
- [x] New public packages, classes, and methods are **documented**. (Strongly consider documenting private classes and methods.)

#### Blockers

Add a checkbox for yourself with your **&#64;name** to this list if you are holding this PR open for review. PR Shepherd, please hold off merging this PR until these are all checked:

- [x] @datalogics-coreyy
- [x] @datalogics-robl

datalogics-kam committed:
FillFormTest: Account for decimal point change in calculatedNumber

- New auto-calculation replaces the stored number with a calculated
  floating point number that is expressed with one decimal place.

datalogics-kam committed:
FillFormTest: Put challenging value into the FDF file

- Put in a badly calculated value to show that the auto recalculation
  accomplishes a visible change.

datalogics-kam committed:
FillForm: Remove explicit call to running calculate scripts.

- They're run automatically when importing the form after PDFJT 4.4.1.

datalogics-devauto commented:
Build failed.

datalogics-devauto commented:
Build failed.

datalogics-devauto commented:
Build failed.

datalogics-devauto commented:
Build failed.

datalogics-kam commented:
Note that tests will fail until [datalogics/pdf-java-toolkit#1007](https://octocat.dlogics.com/datalogics/pdf-java-toolkit/pull/1007) is merged.

datalogics-coreyy commented:
Closing and reopening to kick off CI

datalogics-devauto commented:
Build succeeded.

datalogics-devauto commented:
Build succeeded.

datalogics-coreyy commented on src/main/java/com/datalogics/pdf/samples/forms/FillForm.java:
```diff
@@ -177,8 +177,6 @@ public static void fillAcroformFdf(final PDFDocument pdfDocument, final URL inpu
         fdfService.importForm(fdfDocument);
 
-        // Run calculations on the AcroForm...
-        FormFieldService.getAcroFormFieldManager(pdfDocument).runCalculateScripts();
```
The call to runCalculateScripts() should be removed for XFDF as well

datalogics-kam replied:
Thanks. Turns out I was searching the wrong class for those functions to remove.

datalogics-kam replied:
I took out the calculation for XFDF.

Then, I put a "challenging" value into the XFDF file, and the test output fails in a way that seems to indicate that the calculations didn't happen.

/cc @datalogics-robl

datalogics-robl replied:
@datalogics-kam I understand what went wrong, and I maintain that it is not a problem with the ``XFDFService`` code but with the PDF we're using in the ``FillForm`` sample. To verify what you're trying to prove, we'll need to make substantive changes to the sample and the test.

If both @datalogics-kam and @datalogics-coreyy this will be easier for me to demonstrate in person tomorrow.

datalogics-kam replied:
I think the problem is that the calculation order in the form has the barcodes calculated before the calculatedValue. Calculating everything twice makes it converge, and that's why it worked before removing the manual calculations.

So I can see the mechanism by which the results come out that way. I'm still trying to figure out if it's a problem or not, considering that the spec talks about _all_ calculations being done in the calculation order _every time_ a field value changes. What does that mean for import?

I'm going to go see if I can make Acrobat get this wrong too...because if Acrobat pulls through, I think PDFJT should as well.

datalogics-kam replied:
Fortunately, Acrobat messes this up, too. If you import the XFDF data and save the file, the barcodes are wrong.

datalogics-kam replied:
I'll fix the source PDF.

datalogics-kam committed:
FillFormTest: Put challenging value into the XFDF file

- Put in a badly calculated value to show that the auto recalculation
  accomplishes a visible change.

datalogics-kam committed:
FillForm: Remove explicit call to running calculate scripts for XFDF.

- They're run automatically when importing the form after PDFJT 4.4.1.

datalogics-devauto commented:
Build failed.

datalogics-devauto commented:
Build failed.

datalogics-kam committed:
Fix calculation order in XFDF form-filling example.

- Calculate the calculatedNumber before any of the barcodes, so that the
  barcodes include the correct value as they aggregate up other fields.
- This used to seem to work before when the form data fortuitously was
  the same as the result of calculation. Minor formatting changes in the
  calculation, and deliberate changes to the input file showed up the
  longstanding problem with the sample PDF file.
- Acrobat shows the same misbehavior if you, say, import the form and
  inspect the barcode value, or save the PDF and inspect the values
  stored in the PDF.

datalogics-kam committed:
FillFormTest: Correct id field in dumped XFDF data.

- Changing the example document changed the modified id.

datalogics-kam committed:
Move PDFJT version check to the FillForm class.

- So it's visible for both running and for tests.
- Be able to compare Maven version numbers in main, not just test.

datalogics-devauto commented:
Build failed.

datalogics-kam commented:
This should be happier now.

- Fixed the example form to do the calculations in a better order.
- Got the tests to work with Talkeetna 5/PDFJT 3.

Comprehensive testing could be done with these two commands:

```bash
$ mvn help:active-profiles dependency:tree clean verify
```

```bash
$ mvn -Puse-talkeetna-6 help:active-profiles dependency:tree clean verify
```

(the extra goals merely log the active profiles and the dependencies for visual verification)

datalogics-devauto commented:
Build succeeded.

datalogics-devauto commented:
Build succeeded.

datalogics-devauto commented:
Build succeeded.

datalogics-devauto commented:
Build succeeded.

datalogics-devauto commented:
Build succeeded.

datalogics-coreyy commented on lite/pom.xml:
```diff
@@ -98,7 +98,6 @@
       <groupId>org.apache.maven</groupId>
       <artifactId>maven-artifact</artifactId>
       <version>3.3.9</version>
-      <scope>test</scope>
```
What are these changes in the two poms accomplishing?

datalogics-kam replied:
These change the scope of the ``maven-artifact`` dependency to compile, the default. Since I moved the version check from tests-only to main code, I had to move the dependency over as well.

datalogics-coreyy replied:
Ah I see now. Thanks for clarifying.

datalogics-coreyy commented:
Reviewed and tested locally. All is good for me...

datalogics-robl commented:
Looks good to me as well...
